### PR TITLE
[DataGrid] Add `disableVirtualization` to disable virtualization completely

### LINKED ISSUE #1781: [DataGrid] Allow to disable virtualization
It should be possible to disable row virtualization.

Reasons why you would want to disable row virtualization:

1. It introduces *microstutter* when scrolling, especially on mobile
2. It breaks <kbd>Ctrl</kbd>+<kbd>F</kbd> for users
3. It doesn't provide much benefit if the page size is small

### PR DESCRIPTION
**Pull request by m4theushw**
Preview https://deploy-preview-2326--material-ui-x.netlify.app/components/data-grid/virtualization/#disable-virtualization

Closes #1781

I created a separate hook to not mix concerns and to touch as little as possible the virtualization logic.

- cf7ac773 moves the `scrollToIndexes` and `scroll` to a dedicated hook because they're not tied to the virtualization
- 624352ee better name the hook
- fe60f827 adds a `useGridNoVirtualization` hook to sync the state and handle the scroll when virtualization is off

Preview: https://deploy-preview-2326--material-ui-x.netlify.app/storybook/?path=/story/x-grid-tests-rows--disable-virtualization&globals=measureEnabled:false

### CODE REVIEW

**Comment by oliviertassinari** [comment]
Note that we still have 1. because of how scroll needs to be sync with translate. I think that we should look into dropping this pattern for #1933. We are the only one in the benchmark to not directly rely on the native scroll feature rendering.

**Comment by m4theushw** [comment]
There's a bug in this PR. Filter and try to navigate with the keyboard, it doesn't work.

---

> We are the only one in the benchmark to not directly rely on the native scroll feature rendering.

@oliviertassinari Yeah, I'll investigate if it's easier to use the native scroll instead. Basically, we need to remove the two divs between the window and the rendering zone.

image

> Do we disable column virtualization?

Yes

> Should we add documentation about it in https://material-ui.com/components/data-grid/virtualization/?

I added a small section in https://deploy-preview-2326--material-ui-x.netlify.app/components/data-grid/virtualization/#disable-virtualization

File: docs/src/pages/components/data-grid/virtualization/virtualization.md [hunk 1]
```diff
@@ -33,9 +33,13 @@ By default, 2 columns are rendered outside of the viewport. You can change this
 
 You can disable column virtualization by setting the column buffer to a higher number than the number of rendered columns, e.g. with `columnBuffer={columns.length}` or `columnBuffer={Number.MAX_SAFE_INTEGER}`.
 
-## apiRef [<span class="pro"></span>](https://material-ui.com/store/items/material-ui-pro/)
+## Disable virtualization
 
-{{"demo": "pages/components/data-grid/virtualization/VirtualizationApiNoSnap.js", "bg": "inline", "hideToolbar": true}}
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/virtualization/virtualization.md:38]: Why did you remove this section?

Comment (m4theushw) [reply docs/src/pages/components/data-grid/virtualization/virtualization.md:38]: One of the changes is to move the scroll methods to a separated hook. Since this interface is empty now I removed it. Maybe we should add a page for scrolling with a demo: https://github.com/mui-org/material-ui-x/issues/1103#issuecomment-885695797. In this new page we could document the API available.

**Comment by flaviendelangle** [comment]
Probably also closes #1519

File: packages/grid/_modules_/grid/hooks/features/virtualization/useGridNoVirtualization.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  const visibleColumns = useGridSelector(apiRef, visibleGridColumnsSelector);
+
+  const syncState = React.useCallback(() => {
+    if (!gridState.containerSizes || !windowRef?.current) {
+      return;
+    }
+
+    let firstRowIdx = 0;
+    const { page, pageSize } = paginationState;
+    if (props.pagination && pageSize != null && props.paginationMode === 'client') {
```
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/virtualization/useGridNoVirtualization.ts:30]: `pageSize` can never be null here
See https://github.com/mui-org/material-ui-x/pull/2433/files/0eb2f693687f04990c2b318706cb4923ac25c91c#r695592921

**Review by flaviendelangle — CHANGES_REQUESTED** [review]

**Review by DanailH — APPROVED** [review]
